Various widget layout changes

jdalsem opened the pull request:
## feat(widgets): adding widgets now opens in a lightbox with more info

fixes #9513

![add_widgets_in_lightbox](https://cloud.githubusercontent.com/assets/861833/19808329/ab4cd4ca-9d24-11e6-8534-17dab17563ac.png)
- [x] Write / Update docs
- [x] Test default widgets
- [x] Fix admin css

## feat(dashboard): only show blurb if there are no widgets

fixes #10352

## feat(profile): the profile layout page now uses a 2 column widget layout

![2_col_profile](https://cloud.githubusercontent.com/assets/861833/19811222/c49f54a0-9d30-11e6-8058-b96e70549548.png)

jdalsem committed:
feat(widgets): adding widgets now opens in a lightbox with more info

fixes #9513

jdalsem commented:
Should i separate widgets css in a separate file... so we can use it in frontend en admin? I think YES

I think i will also refactor the rest of the widgets js, so there will be no more `elgg/widgets`...

jdalsem committed:
feat(dashboard): only show blurb if there are no widgets

fixes #10352

jdalsem committed:
feat(profile): the profile layout page now uses a 2 column widget layout

jdalsem commented:
changed some more to make the widget layout make more sense...

I had to change somethings in the dashboard and profile plugin for it to work again... passing content to the widget layout is no longer supported... there is a special 'no_widgets' parameter to echo a string or Closure if no widgets are present.

Dashboard uses the 'no widgets' parameter for the blurb, so you only see the intro text if there are no widgets.

Profile can no longer pass the wrapper view as content, so instead of making the profile details part of the widget layout, i set it on top of the widget layout. Also switched to a 2 column layout (BC with 3 column layout, by failing over widgets that are over the max columns into the last visible one).

jdalsem commented:
see screenshot in the initial post about the profile page changes

hypeJunction commented:
Are widgets columns in the nornal left to right order?

hypeJunction commented:
Let's just add owner block to prpfile pages and not do crazy stuff with user hover.

jdalsem commented:
Did not change widget column order,  so column 1 is still the right most column

jdalsem commented:
U faked the profile screenshot a bit. Was not planning on touching more if the profile wrapper... We could improve in the layout after this change

hypeJunction commented:
> Did not change widget column order, so column 1 is still the right most column

IMO, we should just swap. The only reason it's the other way around is because of the content wrapper that was there.

> Was not planning on touching more if the profile wrapper... We could improve in the layout after this change

Just a suggestion. We can do it later.

hypeJunction commented on mod/profile/views/default/profile/layout.php:
```diff
@@ -1,13 +1,7 @@
-);
-echo elgg_view_layout('widgets', $params);
+echo elgg_view('profile/wrapper');
+echo elgg_view_layout('widgets', ['num_columns' => 2]);
```
Pass owner explicitly, chain from resource view if needed. Enough of this page owner magic.

jdalsem replied:
done

hypeJunction commented on views/default/page/layouts/widgets.php:
```diff
@@ -35,64 +35,82 @@
 
-if (isset($vars['content'])) {
-	echo $vars['content'];
+$widget_class = "elgg-col-1of{$num_columns}";
```
Let's get rid of these and use flexbox.

jdalsem replied:
changing the classes will break to much (in WidgetManager for example)...

hypeJunction commented on views/default/page/layouts/widgets.php:
```diff
@@ -35,64 +35,82 @@
+	
+	$widgets_content = '';
+	foreach ($column_widgets as $widget) {
+		if (!array_key_exists($widget->handler, $widget_types)) {
```
Can this be done in ege* for widgets?

jdalsem replied:
thought about this myself too... not doing this for now... `elgg_get_widget_types` has support for exact match, don't want to burden `elgg_get_widgets` with that logic... letting the layout decide this for now... going to do a follow up to remove registering widgets for 'all' context and will remove the exact match then... makes it easier to do a valid handler check in `elgg_get_widgets`

jdalsem commented:
Swapping the columns would cause a lot of issues... Only way to prevent it is to make some crazy migrate script...

hypeJunction commented on views/default/page/layouts/widgets.php:
```diff
@@ -35,64 +35,82 @@
+		'id' => "elgg-widget-col-{$column_index}",
+		'class' => [
+			"elgg-col-1of{$num_columns}",
+			'elgg-widgets',
```
`elgg-widgets-column` ideally.

jdalsem replied:
not changing this for now

hypeJunction commented on views/default/page/layouts/widgets.php:
```diff
@@ -35,64 +35,82 @@
 elgg_pop_context();
 
-echo elgg_view('graphics/ajax_loader', array('id' => 'elgg-widget-loader'));
+$result .= elgg_view('graphics/ajax_loader', ['id' => 'elgg-widget-loader']);
```
Get rid of this and use spinner.

jdalsem replied:
leaving this for now...

hypeJunction commented on views/default/page/layouts/widgets/add_button.php:
```diff
@@ -2,26 +2,30 @@
+			'maxWidth' => '900px',
+			'maxHeight' => '90%',
+		]),
+		'onclick' => 'require(["page/layouts/widgets/add_panel"])',
```
Make this a menu item and pass with `items` to `elgg_view_menu('title:widgets')` ? Use menu class instead of `elgg-widgets-add-control`

jdalsem replied:
done

hypeJunction commented on views/default/page/layouts/widgets/add_panel.js:
```diff
@@ -0,0 +1,30 @@
+		var $item = $(this).closest('li');
+
+		// if multiple instances not allow, disable this widget type add button
+		var multiple = $item.attr('class').indexOf('elgg-widget-multiple') != -1;
```
`!$item.is('.elgg-widget-multiple')`

jdalsem replied:
done

hypeJunction commented on views/default/page/layouts/widgets/add_panel.js:
```diff
@@ -0,0 +1,30 @@
+			$item.toggleClass('elgg-state-unavailable elgg-state-available');
+		}
+
+		elgg.get($(this).attr('href'), {
```
switch to `elgg/Ajax`

jdalsem replied:
done

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
 
-$widgets = elgg_extract('widgets', $vars);
+if (!elgg_is_xhr()) {
+	echo elgg_echo('ajax:not_is_xhr');
```
log error instead?

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
 <?php
 /**
  * Widget add panel
- *
```
Why remove these?

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
+$exact = elgg_extract('exact_match', $vars);
+$owner_guid = (int) elgg_extract('owner_guid', $vars);
+$owner = get_entity($owner_guid);
+elgg_set_page_owner_guid($owner_guid);
```
Make this view into a properly handled page with `elgg_ajax_gatekeeper()`. Setting page owner in views makes me icky.

jdalsem replied:
moved to a pagehandler

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
+
+	$class[] = $widget_type->multiple ? 'elgg-widget-multiple' : 'elgg-widget-single';
+	
+	$item_content = '<div class="elgg-widgets-add-actions float-alt">';
```
Can we do this without floats please?

hypeJunction replied:
Make it into a menu and float it with CSS if needed. I just don't want any more inline helper class uses. Makes it impossible for other plugin to work with without replacing the view.

jdalsem replied:
moved the float to the css... not making it a menu as it complexes the code needlessly combined with the status text about not being able to add multiple...

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
+	
+	$item_content = '<div class="elgg-widgets-add-actions float-alt">';
+	if (!$widget_type->multiple) {
+		$item_content .= elgg_format_element('span', ['class' => 'elgg-quiet'], elgg_echo('widget:unavailable'));
```
Too long

hypeJunction commented on views/default/page/layouts/widgets/add_panel.php:
```diff
@@ -1,69 +1,87 @@
+	]);
+	$item_content .= '</div>';
+	
+	$item_content .= "<h3>{$widget_type->name}</h3>";
```
Why `h3`? We should really audit our usages of headings, really bad for SEO.

jdalsem replied:
changed to h4 (as module heading is already h3).... but should module heading not be a h2?... other discussion

hypeJunction commented:
> Swapping the columns would cause a lot of issues..

Such as?

jdalsem commented:
Them being in the wrong order... 

Btw... Nice review

hypeJunction commented:
If you ask me, that's the correct order. RTL site can swap them if it needs to.

jeabakker commented on engine/lib/views.php:
```diff
@@ -1901,7 +1901,7 @@ function elgg_views_boot() {
 
 	elgg_register_ajax_view('languages.js');
-
+	
```
?????

jeabakker commented on views/default/admin.css.php:
```diff
@@ -1194,110 +1194,11 @@
-	margin-bottom: 15px;
-}
+<?= elgg_view('elements/widgets.css', $vars) ?>
 
```
what about widget header colors? If the front-end css is used on the backend?

jeabakker replied:
@Elgg/core thoughts

hypeJunction replied:
I don't mind

jdalsem commented:
addressed most of the comments by @hypeJunction 

hoping for a merger... can do follow ups for semi related stuff (that remained the same, but should also be addressed in the future)
